onnxruntime/core/providers/dml/DmlExecutionProvider/src/TensorDesc.cpp

```diff
@@ -315,3 +315,52 @@
     }
     m_bufferTensorDesc.DimensionCount = newDimensionCount;
 }
+
+// Uses dimensionMapping to reorder m_sizes and m_strides to match specific Tensor layout
+void TensorDesc::PermuteDimensions(const std::vector<uint32_t> dimensionMapping, const TensorAxis alignment)
+{
+    EnsureMinimumDimensionCount(dimensionMapping.size(), alignment);
+    InitializeStrides(static_cast<int>(dimensionMapping.size()), alignment);
+    PermuteArray(dimensionMapping, alignment);
+}
+
+// Shuffle m_sizes and m_strides acording to the indexes pointed by dimensionMapping
+void TensorDesc::PermuteArray(const std::vector<uint32_t> dimensionMapping, const TensorAxis alignment)
+{
+    std::vector<uint32_t> temp_sizes(dimensionMapping.size());
+    std::vector<uint32_t> temp_strides(dimensionMapping.size());
+    // Right alignment values are shifted to the end
+    int alignmentOffset = (alignment == TensorAxis::RightAligned) ? MaximumDimensionCount - static_cast<int>(dimensionMapping.size()) : 0;
+
+    for (size_t i = 0; i < dimensionMapping.size(); i++) {
+        temp_sizes[i] = m_sizes[dimensionMapping[i] + alignmentOffset];
+        temp_strides[i] = m_strides[dimensionMapping[i] + alignmentOffset];
+    }
+
+    std::copy(temp_sizes.begin(), temp_sizes.end(), m_sizes + alignmentOffset);
+    std::copy(temp_strides.begin(), temp_strides.end(), m_strides + alignmentOffset);
+    m_bufferTensorDesc.Sizes = m_sizes;
+    m_bufferTensorDesc.Strides = m_strides;
+}
+
+void TensorDesc::InitializeStrides(int dimensionCount, const TensorAxis alignment)
+{
+    ML_CHECK_VALID_ARGUMENT(alignment == TensorAxis::RightAligned || alignment == TensorAxis::LeftAligned);
+    // Right alignment values are shifted to the end
+    int alignmentOffset = (alignment == TensorAxis::RightAligned) ? MaximumDimensionCount - dimensionCount : 0;
+    int index = dimensionCount + alignmentOffset;
+    uint32_t stride = 1;
+    m_strides[index - 1] = 1;
+    for (int i = index - 2; i >= 0; i--)
+    {
+        stride *= m_sizes[i + 1];
+        m_strides[i] = stride;
+    }
+}
+
+void TensorDesc::EnsureMinimumDimensionCount(const size_t dimensionCount, const TensorAxis alignment)
+{
+    // m_sizes and m_strides are arrays of size MaximumDimensionCount
+    assert(MaximumDimensionCount >= dimensionCount);
```

Comment:
>assert(MaximumDimensionCount >= dimensionCount); [](http://example.com/codeflow?start=4&length=48)

Should be an `if` rather than an `assert`. It's okay if the dimensions are less than the new dimension count, as this function will ensure (before it returns) that they *become* at least the minimum dimension count. This is actually a desirable utility, as there are many cases in ONNX where an incoming tensor is only 1D, and we want to expand it to ND to match the size of another tensor like the input.